[SPMD] Allow IterGraph support a more general subgraph movement

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98360

Resubmit D44444398 due to the  merging conflict.

The original assumption of IterGraph is very restrictive and only allow users to move a subgraph that only one node has the input from external nodes. This PR fixes the limitation.

Differential Revision: [D44689730](https://our.internmc.facebook.com/intern/diff/D44689730/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44689730/)!

fegin committed:
[SPMD] Allow IterGraph support a more general subgraph movement

Resubmit D44444398 due to the  merging conflict.

The original assumption of IterGraph is very restrictive and only allow users to move a subgraph that only one node has the input from external nodes. This PR fixes the limitation.

Differential Revision: [D44689730](https://our.internmc.facebook.com/intern/diff/D44689730/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44689730/)!

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Allow IterGraph support a more general subgraph movement"


Resubmit D44444398 due to the  merging conflict.

The original assumption of IterGraph is very restrictive and only allow users to move a subgraph that only one node has the input from external nodes. This PR fixes the limitation.

Differential Revision: [D44689730](https://our.internmc.facebook.com/intern/diff/D44689730/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44689730/)!

[ghstack-poisoned]

lessw2020 commented on torch/distributed/_spmd/iter_graph_module.py:
```diff
@@ -78,174 +83,261 @@ def _lookup_node(self, node: fx.Node, graph: fx.Graph) -> Optional[fx.Node]:
-                "The nodes for move_to_next_iter_before must form a SESE "
-                "subgraph. The output of this subgraph should be the output "
-                " of the whole graph."
+                "The target nodes for ``move_to_next_iter_before`` must "
```
+1 on this  very clear error reason breakout for when it is rejected.

lessw2020 reviewed (approved):
lgtm!

fegin committed:
Update on "[SPMD] Allow IterGraph support a more general subgraph movement"


Resubmit D44444398 due to the  merging conflict.

The original assumption of IterGraph is very restrictive and only allow users to move a subgraph that only one node has the input from external nodes. This PR fixes the limitation.

Differential Revision: [D44689730](https://our.internmc.facebook.com/intern/diff/D44689730/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44689730/)!

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Allow IterGraph support a more general subgraph movement"


Resubmit D44444398 due to the  merging conflict.

The original assumption of IterGraph is very restrictive and only allow users to move a subgraph that only one node has the input from external nodes. This PR fixes the limitation.

Differential Revision: [D44689730](https://our.internmc.facebook.com/intern/diff/D44689730/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44689730/)!

[ghstack-poisoned]
